lib/src/menus/multi_level_menus.dart

```diff
@@ -0,0 +1,111 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+class MultiLevelMenuController with ChangeNotifier {
+  MenuPath? get activePath => _activePath;
+  MenuPath? _activePath;
+
+  bool isActive(MenuPath menuPath) {
+    if (_activePath == null) {
+      return false;
+    }
+
+    return menuPath.isSubPathOf(_activePath!);
+  }
+
+  void activate(MenuPath menuPath) {
+    if (_visiblePath == null || !menuPath.isSubPathOf(_visiblePath!)) {
+      debugPrint("WARNING: Can't activate path because it's not a sub-path of the visible path.");
+      debugPrint(" - desired active path: ${menuPath.serialize()}");
+      debugPrint(" - visible path: ${_visiblePath?.serialize()}");
+      return;
+    }
+
+    _activePath = menuPath;
+    notifyListeners();
+  }
+
+  MenuPath? get visiblePath => _visiblePath;
+  MenuPath? _visiblePath;
+
+  bool isVisible(MenuPath menuPath) {
+    if (_visiblePath == null) {
+      return false;
+    }
+
+    return menuPath.isSubPathOf(_visiblePath!);
+  }
+
+  void show(MenuPath menuPath) {
+    _visiblePath = menuPath;
+    notifyListeners();
+  }
+
+  void close() {
+    _visiblePath = null;
+    notifyListeners();
+  }
+}
+
+class MenuPath {
+  const MenuPath(this._path);
+
+  List<String> get components => List.from(_path);
```

Comment:
We could use `UnmodifiableListView` here instead of `List.from` to avoid copying the list. It might not not make much of a difference though.

Reply:
I think we can leave as-is for now. If it becomes a problem then we can change it.